### DISCUSSION TITLE
[SSL: SHUTDOWN_WHILE_IN_INIT] shutdown while in init (_ssl.c:2685)

### DISCUSSION THREAD

**Discussion by rsome93** [openssl/openssl#25802] [Q&A]
I am trying to connect to ftp server from aws lambda and retrieve python files using ftplib library. But even though it establishes the connection it is sometimes failing  due to [SSL: SHUTDOWN_WHILE_IN_INIT] shutdown while in init (_ssl.c:2685)
 Any suggestions?

**Comment by tom-cosgrove-arm** [comment]
We usually prefer questions like this to be asked using GitHub discussions (the "New Issue" screen actually says "Please use Q&A in Discussions instead")

**Comment by jarach** [comment]
@rsome93 
I have found working solution here [getting-a-oserror-when-trying-to-list-ftp-directories-in-python](https://stackoverflow.com/questions/46633536/getting-a-oserror-when-trying-to-list-ftp-directories-in-python). It looks like the problem is on FTPS server side. ftplib developers could implement this workaround...

Below you will find working for me a piece of code:
```
from ftplib import FTP_TLS,FTP
import ssl
from ssl import SSLSocket

class ReusedSslSocket(SSLSocket):
    def unwrap(self):
        pass
class FTP_TLS_IgnoreHost(FTP_TLS):
    def makepasv(self):
        _, port = super().makepasv()
        return self.host, port
    """Explicit FTPS, with shared TLS session"""
    def ntransfercmd(self, cmd, rest=None):
        conn, size = FTP.ntransfercmd(self, cmd, rest)
        if self._prot_p:
            conn = self.context.wrap_socket(conn,
                      server_hostname=self.host,
                      session=self.sock.session)  # reuses TLS session            
            conn.__class__ = ReusedSslSocket  # we should not close reused ssl socket when file transfers finish
        return conn, size

ftp = FTP_TLS_IgnoreHost('x.x.x.x', timeout=5)
ftp.ssl_version = ssl.PROTOCOL_TLS
ftp.auth()
ftp.login('user', 'password')
ftp.prot_p()
ftp.set_pasv(True)
pwd = ftp.pwd()
print(pwd)
print(ftp.nlst())
ftp.close()
```
